src/kv/RocksDBStore.cc

```diff
@@ -495,14 +495,381 @@ int RocksDBStore::load_rocksdb_options(bool create_if_missing, rocksdb::Options&
+    if (!b) {
+      dout(1) << __func__ << " bad sharding: " << dendl;
+      dout(1) << __func__ << sharding_text << dendl;
+      dout(1) << __func__ << std::string(error_position - &sharding_text[0], ' ') << "^" << error_msg << dendl;
+      return -EINVAL;
+    }
+    r = create_shards(opt, sharding_def);
+    if (r != 0 ) {
+      return r;
+    }
+    opt.env->CreateDir(sharding_def_dir);
+    status = rocksdb::WriteStringToFile(opt.env, sharding_text,
+					sharding_def_file, true);
+    if (!status.ok()) {
+      derr << __func__ << " cannot write to " << sharding_def_file << dendl;
+      return -EIO;
+    }
+  } else {
+    opt.env->DeleteFile(sharding_def_file);
+  }
   return 0;
 }
 
+int RocksDBStore::verify_sharding(const rocksdb::Options& opt,
+				  const std::string& sharding_text,
+				  std::vector<rocksdb::ColumnFamilyDescriptor>& existing_cfs,
+				  std::vector<std::pair<size_t, RocksDBStore::ColumnFamily> >& existing_cfs_shard,
+				  std::vector<rocksdb::ColumnFamilyDescriptor>& missing_cfs,
+				  std::vector<std::pair<size_t, RocksDBStore::ColumnFamily> >& missing_cfs_shard)
+{
+  rocksdb::Status status;
+  std::string stored_sharding_text;
+  status = opt.env->FileExists(sharding_def_file);
+  if (status.ok()) {
+    status = rocksdb::ReadFileToString(opt.env,
+				       sharding_def_file,
+				       &stored_sharding_text);
+    if(!status.ok()) {
+      derr << __func__ << " cannot read from " << sharding_def_file << dendl;
+      return -EIO;
+    }
+  } else {
+    //no "sharding_def" present
+  }
+  //check if sharding_def matches stored_sharding_def
+  std::vector<ColumnFamily> sharding_def;
+  std::vector<ColumnFamily> stored_sharding_def;
+  if (!sharding_text.empty()) {
+    parse_sharding_def(sharding_text, sharding_def);
+  } else {
+    //if sharding requested is empty, assume that it agrees with stored
+    //this is necessary for ceph-bluestore-tool fsck
+    parse_sharding_def(stored_sharding_text, sharding_def);
+  }
+  parse_sharding_def(stored_sharding_text, stored_sharding_def);
+
+  std::sort(sharding_def.begin(), sharding_def.end(),
+	    [](ColumnFamily& a, ColumnFamily& b) { return a.name < b.name; } );
+  std::sort(stored_sharding_def.begin(), stored_sharding_def.end(),
+	    [](ColumnFamily& a, ColumnFamily& b) { return a.name < b.name; } );
+
+  bool match = true;
+  if (sharding_def.size() != stored_sharding_def.size()) {
+    match = false;
+  } else {
+    for (size_t i = 0; i < sharding_def.size(); i++) {
+      auto& a = sharding_def[i];
+      auto& b = stored_sharding_def[i];
+      if ( (a.name != b.name) ||
+	   (a.shard_cnt != b.shard_cnt) ||
+	   (a.hash_l != b.hash_l) ||
+	   (a.hash_h != b.hash_h) ) {
+	match = false;
+	break;
+      }
+    }
+  }
+  if (!match) {
+    derr << __func__ << " mismatch on sharding. requested = " << sharding_def
+	 << " stored = " << stored_sharding_def << dendl;
+    return -EIO;
```

Comment:
when running the cephadm upgrade tasks from octopus, I'm getting:

https://tracker.ceph.com/issues/45335